header/p2p/session.go

```diff
@@ -0,0 +1,226 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/header"
+	p2p_pb "github.com/celestiaorg/celestia-node/header/p2p/pb"
+)
+
+// TODO(@vgonkivs): make it configurable
+var (
+	// headersPerPeer is a maximum amount of headers that will be requested per peer.
+	headersPerPeer uint64 = 64
+	// sessionDuration is an amount of time after which session will be closed.
+	sessionDuration = time.Minute * 2
+)
+
+// session aims to divide a range of headers
+// into several smaller requests among different peers.
+type session struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	host       host.Host
+	protocolID protocol.ID
+	// peerTracker contains discovered peers with records that describes their activity.
+	queue *peerQueue
+
+	reqCh chan *p2p_pb.ExtendedHeaderRequest
+	errCh chan error
+}
+
+func newSession(ctx context.Context, h host.Host, peerTracker []*peerStat, protocolID protocol.ID) *session {
+	ctx, cancel := context.WithTimeout(ctx, sessionDuration)
+	return &session{
+		ctx:        ctx,
+		cancel:     cancel,
+		protocolID: protocolID,
+		host:       h,
+		queue:      newPeerQueue(peerTracker),
+		errCh:      make(chan error),
+	}
+}
+
+// GetRangeByHeight requests headers from different peers.
+func (s *session) getRangeByHeight(ctx context.Context, from, amount uint64) ([]*header.ExtendedHeader, error) {
+	log.Debugw("requesting headers", "from", from, "to", from+amount)
+	requests := prepareRequests(from, amount, headersPerPeer)
+	result := make(chan []*header.ExtendedHeader, len(requests))
+	headers := make([]*header.ExtendedHeader, 0, amount)
+	s.reqCh = make(chan *p2p_pb.ExtendedHeaderRequest, len(requests))
+
+	go s.handleOutgoingRequests(ctx, result)
+	for _, req := range requests {
+		s.reqCh <- req
+	}
+	for i := 0; i < cap(result); i++ {
+		select {
+		case <-s.ctx.Done():
+			return nil, errors.New("header/p2p: session was stopped")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-s.errCh:
+			return nil, err
+		case res := <-result:
+			headers = append(headers, res...)
+		}
+	}
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Height < headers[j].Height
+	})
+	return headers, nil
+}
```

Comment:
```suggestion
	log.Debugw("requesting headers", "from", from, "to", from+amount)

	requests := prepareRequests(from, amount, headersPerPeer)
	s.reqCh = make(chan *p2p_pb.ExtendedHeaderRequest, len(requests))
	result := make(chan []*header.ExtendedHeader, len(requests))

	// spawn requests
	go s.handleOutgoingRequests(ctx, result)
	for _, req := range requests {
		s.reqCh <- req
	}

	headers := make([]*header.ExtendedHeader, 0, amount)
	for i := 0; i < cap(result); i++ {
		select {
		case <-s.ctx.Done():
			return nil, errors.New("header/p2p: session was stopped")
		case <-ctx.Done():
			return nil, ctx.Err()
		case err := <-s.errCh:
			return nil, err
		case res := <-result:
			headers = append(headers, res...)
		}
	}

	sort.Slice(headers, func(i, j int) bool {
		return headers[i].Height < headers[j].Height
	})
	return headers, nil
```

I think this reads better. We should ensure to make larger functions like this more readable by grouping related things together